src/main/java/com/adroll/cantor/HLLWritable.java

```diff
@@ -254,23 +253,28 @@ public void readFields(DataInput in) throws IOException {
         M = new HLLByteArray(m);
       } else {
         int m = (int)Math.pow(2, p);
-        M = new HLLByteArray(m);
-        for(int i = 0; i < m; i++) {
-          M.put(i, in.readByte());
-        }
+        byte[] t = new byte[m];
+        in.readFully(t, 0, m);
+        M = new HLLByteArray(t);
       }
       minhash = new long[s];
 
-      for(int i = 0; i < s; i++) {
-        long x = in.readLong();
-        minhash[i] = x;
-        /**
-         * If p was negative, M is empty and we need to re-populate
-         * If p was positive and we read M, this won't change anything since it's just max
-         */
-        int idx = (int)(x >>> (64 - p));
-        long w = x << p;
-        M.put(idx, M.get(idx) > Long.numberOfLeadingZeros(w) + 1 ? M.get(idx) : (byte)(Long.numberOfLeadingZeros(w) + 1));
+      byte[] data = new byte[s*8];
+      in.readFully(data, 0, s*8);
+      ByteBuffer.wrap(data).asLongBuffer().get(minhash);
+
+      long x;
+      if (s < k) {
```

Comment:
why this condition? I'm interested for my own edification but also a comment here would prevent any future confusion.